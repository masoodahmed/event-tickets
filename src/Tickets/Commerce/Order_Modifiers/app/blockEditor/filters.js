import { addFilter, addAction } from '@wordpress/hooks';
