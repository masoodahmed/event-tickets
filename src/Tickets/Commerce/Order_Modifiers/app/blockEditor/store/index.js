/**
 * External dependencies
 */
import { createReduxStore, register } from '@wordpress/data';

/**
 * Internal dependencies
 */
import { getTicketIdFromCommonStore } from './common-store-bridge';
import { defaultState } from './default-state';
import { controls } from './controls';
// import {actions} from './actions';
// import {reducer} from './reducer';
// import {selectors} from './selectors';
import { localizedData } from './localized-data';

const storeName = 'tec-tickets-fees';

const reducer = ( state = defaultState, action ) => {
	switch ( action.type ) {
		case 'SET_TICKET_FEES':
			return {
				...state,
				ticketFees: action.ticketFees,
			};

		case 'SET_AUTOMATIC_FEES':
			return {
				...state,
				feesAutomatic: action.feesAutomatic,
			};

		case 'SET_AVAILABLE_FEES':
			return {
				...state,
				feesAvailable: action.feesAvailable,
			};

		default:
			return state;
	}
};

const actions = {
	// Actions that have to do with fees for a ticket.
	setTicketFees( clientId, feesSelected ) {
		return {
			type: 'SET_TICKET_FEES',
			clientId,
			feesSelected
		};
	},

	// Actions that have to do with managing fees separate from tickets.
	setAutomaticFees( feesAutomatic ) {
		return {
			type: 'SET_AUTOMATIC_FEES',
			feesAutomatic,
		};
	},

	setAvailableFees( feesAvailable ) {
		return {
			type: 'SET_AVAILABLE_FEES',
			feesAvailable,
		};
	},

	fetchFeesFromAPI() {
		return {
			type: 'FETCH_FEES_FROM_API',
		};
	},
};

const selectors = {


	getAutomaticFees( state ) {
		return state.feesAutomatic;
	},

	getAvailableFees( state ) {
		return state.feesAvailable;
	},
};

const resolvers = {
	* getFees() {
		const allFees = yield actions.fetchFeesFromAPI();
	},
};


// @see: https://developer.wordpress.org/block-editor/reference-guides/packages/packages-data/
const store = createReduxStore( storeName, {
	reducer: reducer,
	actions: actions,
	selectors: selectors,
	controls: controls,
	resolvers: resolvers,
} );

register( store );

export { store, storeName };
