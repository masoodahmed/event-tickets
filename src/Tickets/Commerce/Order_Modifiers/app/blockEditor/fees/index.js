/**
 * External dependencies
 */
import classNames from 'classnames';
import { LabeledItem, } from '@moderntribe/common/elements';
import { useSelect, useDispatch, } from '@wordpress/data';
import { useCallback, useState } from '@wordpress/element';
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { setTicketHasChangesInCommonStore } from '../store/common-store-bridge';
import { mapFeeToItem } from './map-fee-object';
import AddFee from './add-fee';
import SelectFee from './select-fee';
import './style.pcss';

// The name of the store for fees.
const storeName = 'tec-tickets-fees';

/**
 * The fees section component for the ticket editor.
 *
 * The fees section needs logic to handle the selection of fees. The default
 * view is to show the "+ Add fee" button. Any previously-selected fees should
 * be displayed as checked above the button.
 *
 * When the "+ Add fee" button is clicked, the user should be able to select
 * from a list of available fees. They can then confirm the selection with the
 * "Add fee" button, or cancel with the "Cancel" button.
 *
 * The selected fees should be displayed as checked above the "+ Add fee" button.
 * Once a fee has been displayed, it should not be hidden even if it is un-checked
 * again. It should only be hidden after it is unchecked AND the page has been
 * reloaded.
 *
 * @since TBD
 *
 * @param {string} props.clientId The client ID of the ticket.
 * @return {JSX.Element}
 * @constructor
 */
function FeesSection( props ) {
	// Extract the clientId from the props.
	const { clientId } = props;

	// Get the available and automatic fees from the store.
	const {
		feesAvailable,
		feesAutomatic,
	} = useSelect(
		( select ) => {
			return {
				feesAvailable: select( storeName ).getAvailableFees(),
				feesAutomatic: select( storeName ).getAutomaticFees(),
			};
		},
		[]
	);

	const hasAutomaticFees = feesAutomatic.length > 0;
	const hasAvailableFees = feesAvailable.length > 0;
	const hasItemsToDisplay = hasAutomaticFees || hasAvailableFees;

	// Set up the state for the selected fees.
	const feesSelected = useSelect(
		( select ) => select( storeName ).getSelectedFees( clientId ),
		[ clientId ]
	);

	const feeIdSelectedMap = {};
	const displayedFees = [];

	// Initialize the selected fees map with the available fees.
	feesAvailable.forEach( ( fee ) => {
		feeIdSelectedMap[ fee.id ] = false;
	} );

	// Set the selected fees to true.
	feesSelected.forEach( ( feeId ) => {
		feeIdSelectedMap[ feeId ] = true;
		let fee = feesAvailable.find( ( fee ) => fee.id === feeId );
		if ( fee ) {
			displayedFees.push( fee );
		}
	} );

	// Set up the state for the selected fees.
	const [ checkedFees, setCheckedFees ] = useState( feeIdSelectedMap );

	// Set up the dispatch functions for adding and removing fees.
	const { addFeeToTicket, removeFeeFromTicket } = useDispatch( storeName );

	/**
	 * Handles the change event for the selected fees.
	 *
	 * @param {event} event The change event.
	 */
	const onSelectedFeesChange = useCallback(
		( event ) => {
			const feeId = Number.parseInt( event.target.value );
			const isChecked = ! checkedFees[ feeId ];

			if ( isChecked ) {
				addFeeToTicket( clientId, feeId );
			} else {
				removeFeeFromTicket( clientId, feeId );
			}

			setCheckedFees( {
				...checkedFees,
				[ feeId ]: isChecked,
			} );

			setTicketHasChangesInCommonStore( clientId );
		},
		[ clientId, checkedFees ]
	);

	// Set up the state for the fee selection.
	const [ isSelectingFee, setIsSelectingFee ] = useState( false );
	const onAddFeeClick = useCallback(
		() => {
			setIsSelectingFee( true );
		},
		[ clientId ]
	);

	// Set up the functions for the fee selection.
	const onCancelFeeSelect = useCallback(
		() => {
			setIsSelectingFee( false );
		},
		[ clientId ]
	);

	const onConfirmFeeSelect = useCallback(
		( feeId ) => {
			// We're done selecting a fee.
			setIsSelectingFee( false );

			// Update the list of checked fees.
			setCheckedFees( {
				...checkedFees,
				[ feeId ]: true,
			} );

			// Dispatch the action to add the fee to the ticket.
			addFeeToTicket( clientId, feeId );
		},
		[ clientId, checkedFees, feesAvailable ]
	)

	// Set up the fees that are available to be selected.
	const selectableFees = feesAvailable.filter( ( fee ) => {
		// The fee should not be selectable if it's already in the display fees.
		return ! displayedFees.some( ( displayedFee ) => displayedFee.id === fee.id );
	} );

	return (
		<div
			className={ classNames(
				'tribe-editor__ticket__fees',
				'tribe-editor__ticket__content-row',
				'tribe-editor__ticket__content-row--fees'
			) }
		>
			<LabeledItem
				className={ classNames(
					'tribe-editor__labeled-item',
					'tribe-editor__ticket__active-fees-label'
				) }
				label={ __( 'Ticket Fees', 'event-tickets' ) }
			/>

			<div className="tribe-editor__ticket__order_modifier_fees">

				{ hasAutomaticFees ? (
					feesAutomatic.map( ( fee ) => mapFeeToItem( {
						isDisabled: true,
						isChecked: true,
						fee: fee,
						clientId: clientId,
					} ) )
				) : null }

				{ displayedFees.length > 0 ? (
					displayedFees.map( ( fee ) => mapFeeToItem( {
						isDisabled: false,
						onChange: onSelectedFeesChange,
						isChecked: checkedFees[ fee.id ],
						fee: fee,
						clientId: clientId,
					} ) )
				) : null }

				{ isSelectingFee
					? <SelectFee
						feesAvailable={ selectableFees }
						onCancel={ onCancelFeeSelect }
						onConfirm={ onConfirmFeeSelect }
					/>
					: <AddFee onClick={ onAddFeeClick }/> }

				{ ! hasItemsToDisplay ? (
					<p>{ __( 'No available fees.', 'event-tickets' ) }</p>
				) : null }
			</div>
		</div>
	);
}

export default FeesSection;
