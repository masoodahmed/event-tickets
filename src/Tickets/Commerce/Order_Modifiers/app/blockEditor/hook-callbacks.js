import { storeName } from './store';
import { select, dispatch } from '@wordpress/data';

/**
 * Filters the body details of the ticket to add the seating details.
 *
 * @since TBD
 *
 * @param {Object} body     The body of the request.
 * @param {string} clientId The client ID of the ticket block.
 *
 * @return {Object} The body of the request with the seating details.
 */
export const filterSetBodyDetails = ( body, clientId ) => {
	const feesSelected = select( storeName ).getSelectedFees( clientId );

	body.append( 'ticket[fees][selected_fees]', feesSelected );

	return body;
};
