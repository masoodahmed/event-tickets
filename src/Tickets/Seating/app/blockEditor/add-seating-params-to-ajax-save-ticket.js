import { storeName } from './store';
import { select, dispatch } from '@wordpress/data';

/**
 * Filters the body details of the ticket to add the seating details.
 *
 * @since TBD
 *
 * @param {Object} body     The body of the request.
 * @param {string} clientId The client ID of the ticket block.
 *
 * @return {Object} The body of the request with the seating details.
 */
export const filterSetBodyDetails = (body, clientId) => {
	const layoutId = select(storeName).getCurrentLayoutId();
	if (!layoutId) {
		return body;
	}

	const seatType = select(storeName).getTicketSeatType(clientId);
	const eventCapacity = select(storeName).getEventCapacity();
	body.append('ticket[seating][enabled]', seatType ? '1' : '0');
	body.append('ticket[seating][seatType]', seatType ? seatType : '');
	body.append('ticket[seating][layoutId]', layoutId);
	body.append('ticket[event_capacity]', eventCapacity);

	// On first save of a ticket, lock the Layout.
	dispatch(storeName).setIsLayoutLocked(true);

	return body;
};
