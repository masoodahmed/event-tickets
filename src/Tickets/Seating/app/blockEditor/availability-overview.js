import { storeName } from './store';
import { select } from '@wordpress/data';

/**
 * Modifies the properties mapped from the state for the Availability component to conform
 * to the Assigned Seating feature.
 *
 * @since TBD
 *
 * @param {Object} mappedProps           The properties mapped from the state for the Availability component.
 * @param {number} mappedProps.total     The total capacity.
 * @param {number} mappedProps.available The available capacity.
 */
export const filterSeatedTicketsAvailabilityMappedProps = (mappedProps) => {
	const store = select(storeName);
	const hasSeats = store.isUsingAssignedSeating();
	const layoutLocked = store.isLayoutLocked();

	if (!(hasSeats && layoutLocked)) {
		return mappedProps;
	}

	const layoutId = store.getCurrentLayoutId();
	if (!layoutId) {
		return mappedProps;
	}

	const seatTypes = store.getSeatTypesForLayout(layoutId, true);
	const activeSeatsByClient = Object.values(store.getSeatTypesByClientID());
	const activeSeatsByPost = Object.values(store.getSeatTypesByPostID());
	const activeSeatTypes =
		activeSeatsByPost.length > activeSeatsByClient.length
			? activeSeatsByPost
			: activeSeatsByClient;

	const activeSeatTypesFiltered = activeSeatTypes.filter(
		(value, index, array) => array.indexOf(value) === index
	);

	const activeSeatTypeTotalCapacity = activeSeatTypesFiltered.reduce(
		(sum, type) => sum + parseInt(seatTypes[type] ? seatTypes[type].seats : 0),
		0
	);

	const seatTypeTotalCapacity = Object.values(seatTypes).reduce(
		(sum, { seats }) => sum + parseInt(seats),
		0
	);

	const soldAndPending = Math.abs(
		parseInt(mappedProps?.total || 0) -
			parseInt(mappedProps?.available || 0)
	);

	return {
		total: seatTypeTotalCapacity,
		available: Math.abs(activeSeatTypeTotalCapacity - soldAndPending),
	};
};
