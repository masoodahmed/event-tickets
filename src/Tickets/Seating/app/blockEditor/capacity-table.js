import {select} from "@wordpress/data";
import { storeName } from './store';
import {addFilter} from "@wordpress/hooks";
import { getLocalizedString } from '@tec/tickets/seating/utils';

export function filterCapacityTableMappedProps(mappedProps) {
	const hasSeats = select(storeName).isUsingAssignedSeating();
	const layoutLocked = select(storeName).isLayoutLocked();

	if ( ! hasSeats || ! layoutLocked ) {
		return mappedProps;
	}

	let layoutId  = select(storeName).getCurrentLayoutId();
	if ( ! layoutId ) {
		return mappedProps;
	}

	let seatTypes = select(storeName).getSeatTypesForLayout(layoutId, true);
	let activeSeatTypes = Object.values( select(storeName).getSeatTypesByPostID() );

	mappedProps.rowsAfter = mappedProps.rowsAfter || [];
	const seatTypeLabels = activeSeatTypes.map( type => seatTypes[type].name );
	const seatTypeTotalCapacity = activeSeatTypes.reduce( ( sum, type ) => sum + parseInt(seatTypes[type].seats), 0 );
	mappedProps.rowsAfter.push({
		label: getLocalizedString( 'seats-row-label', 'capacity-table' ),
		items: seatTypeLabels ? `(${seatTypeLabels})` : '',
		right: String(seatTypeTotalCapacity),
	});

	mappedProps.totalCapacity  = ( mappedProps.totalCapacity - mappedProps.sharedCapacity ) + seatTypeTotalCapacity;
	mappedProps.sharedCapacity = '';
	mappedProps.sharedTicketItems = '';

	return mappedProps;
}

addFilter(
	'tec.tickets.blocks.Tickets.CapacityTable.mappedProps',
	'tec.tickets.flexibleTickets',
	filterCapacityTableMappedProps
);
