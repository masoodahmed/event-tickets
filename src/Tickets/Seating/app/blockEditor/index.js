import './store';
import './filters.js';
import './capacity-table';
