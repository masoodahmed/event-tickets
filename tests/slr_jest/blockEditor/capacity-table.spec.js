import { select } from '@wordpress/data';
import { applyFilters } from '@wordpress/hooks';
import { filterCapacityTableMappedProps } from '@tec/tickets/seating/blockEditor/capacity-table';

jest.mock('@wordpress/data', () => ({
	select: jest.fn(),
	createReduxStore: jest.fn(),
	register: jest.fn(),
}));

describe('filterCapacityTableMappedProps', () => {
	beforeEach(() => {
		jest.resetModules();
		jest.resetAllMocks();
	});

	afterEach(() => {
		jest.resetModules();
		jest.resetAllMocks();
	});

	it('returns original mappedProps if hasSeats is false', () => {
		select.mockReturnValue({ isUsingAssignedSeating: () => false, isLayoutLocked: () => true });
		const mappedProps = { foo: 'bar' };
		const newMappedProps = applyFilters('tec.tickets.blocks.Tickets.CapacityTable.mappedProps', mappedProps);
		expect(newMappedProps).toEqual(mappedProps);
	});

	it('returns original mappedProps if layoutLocked is false', () => {
		select.mockReturnValue({ isUsingAssignedSeating: () => true, isLayoutLocked: () => false });
		const mappedProps = { foo: 'bar' };
		const newMappedProps = applyFilters('tec.tickets.blocks.Tickets.CapacityTable.mappedProps', mappedProps);
		expect(newMappedProps).toEqual(mappedProps);
	});

	it('returns original mappedProps if layoutId is null', () => {
		select.mockReturnValue({ isUsingAssignedSeating: () => true, isLayoutLocked: () => true, getCurrentLayoutId: () => null });
		const mappedProps = { foo: 'bar' };
		const newMappedProps = applyFilters('tec.tickets.blocks.Tickets.CapacityTable.mappedProps', mappedProps);
		expect(newMappedProps).toEqual(mappedProps);
	});

	it('adds rowsAfter and updates totalCapacity if hasSeats and layoutLocked are true', () => {
		const seatTypes = {
			'vip': { name: 'VIP', seats: 20 },
			'general-admission': { name: 'General Admission', seats: 20 },
		};
		const activeSeatTypes = {
			99: 'vip',
		};
		select.mockReturnValue({
			isUsingAssignedSeating: () => true,
			isLayoutLocked: () => true,
			getCurrentLayoutId: () => 'layout-id',
			getSeatTypesForLayout: () => seatTypes,
			getSeatTypesByPostID: () => activeSeatTypes,
		});

		const mappedProps = {
			rowsAfter: [],
			independentCapacity: 20,
			totalCapacity: 40,
			sharedCapacity: 20,
			sharedTicketItems: 'items',
		};

		const result = applyFilters('tec.tickets.blocks.Tickets.CapacityTable.mappedProps', mappedProps);
		expect(result).toMatchSnapshot();
	});

	it('adds rowsAfter and updates totalCapacity if has multiple seat types', () => {
		const seatTypes = {
			'vip': { name: 'VIP', seats: 20 },
			'general-admission': { name: 'General Admission', seats: 20 },
		};
		const activeSeatTypes = {
			99: 'vip',
			100: 'general-admission',
		};
		select.mockReturnValue({
			isUsingAssignedSeating: () => true,
			isLayoutLocked: () => true,
			getCurrentLayoutId: () => 'layout-id',
			getSeatTypesForLayout: () => seatTypes,
			getSeatTypesByPostID: () => activeSeatTypes,
		});

		const mappedProps = {
			rowsAfter: [],
			independentCapacity: 20,
			totalCapacity: 60,
			sharedCapacity: 40,
			sharedTicketItems: 'items',
		};

		const result = applyFilters('tec.tickets.blocks.Tickets.CapacityTable.mappedProps', mappedProps);
		expect(result).toMatchSnapshot();
	});
});