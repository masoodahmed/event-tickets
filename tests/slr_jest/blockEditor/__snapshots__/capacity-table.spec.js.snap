// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterCapacityTableMappedProps adds rowsAfter and updates totalCapacity if has multiple seat types 1`] = `
Object {
  "independentCapacity": 20,
  "rowsAfter": Array [
    Object {
      "items": "(VIP,General Admission)",
      "label": "Assigned seating",
      "right": "40",
    },
  ],
  "sharedCapacity": "",
  "sharedTicketItems": "",
  "totalCapacity": 60,
}
`;

exports[`filterCapacityTableMappedProps adds rowsAfter and updates totalCapacity if hasSeats and layoutLocked are true 1`] = `
Object {
  "independentCapacity": 20,
  "rowsAfter": Array [
    Object {
      "items": "(VIP)",
      "label": "Assigned seating",
      "right": "20",
    },
  ],
  "sharedCapacity": "",
  "sharedTicketItems": "",
  "totalCapacity": 40,
}
`;
